XD-3370: FTP Sink

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/XD-3370

Refactor common source/sink properties for remote file source/sink.

Refactor test servers for use in both source and sink and to properly
use JUnit `TemporaryFolder`.

garyrussell committed:
XD-3370: FTP Sink

JIRA: https://jira.spring.io/browse/XD-3370

Refactor common source/sink properties for remote file source/sink.

Refactor test servers for use in both source and sink and to properly
use JUnit `TemporaryFolder`.

artembilan commented on ftp-sink/pom.xml:
```diff
@@ -0,0 +1,53 @@
+		<dependency>
+			<groupId>org.springframework.integration</groupId>
+			<artifactId>spring-integration-ftp</artifactId>
+			<version>${spring-integration.version}</version>
```
Isn't version provided by the `parent`?
I guess yes, because you don't use `<version>` for the `org.springframework.integration`

garyrussell replied:
To be honest, I just copied the ftp source pom - I didn't pay much attention to it (given my maven aversion :smile: ).

artembilan commented on ftp-sink/pom.xml:
```diff
@@ -0,0 +1,53 @@
+		<dependency>
+			<groupId>org.apache.ftpserver</groupId>
+			<artifactId>ftpserver-core</artifactId>
+		</dependency>
```
Why is without `<scope>test</scope>` ?

garyrussell committed:
POM Polishing

Test scope; remove sshd from ftp modules and ftp from sftp module.

Fix some comments.

garyrussell commented:
Pushed some POM polish.

artembilan commented on common/spring-cloud-stream-modules-common-configuration/src/main/java/org/springframework/cloud/stream/module/file/remote/AbstractRemoteFileProperties.java:
```diff
@@ -0,0 +1,68 @@
+ * @author Gary Russell
+ *
+ */
+public abstract class AbstractRemoteFileProperties {
```
I like more when we mark them all with the `@ConfigurationProperties`, don't inherit and just specify in the `EnableConfigurationProperties`.
See the #46 for similar problem around the common properties.

I won't mind if it doesn't matter though...

artembilan commented on ftp-sink/src/test/java/org/springframework/cloud/stream/module/ftp/FtpSinkIntegrationTests.java:
```diff
@@ -0,0 +1,90 @@
+		properties.put("port", port);
+		properties.put("mode", "FAIL");
+		properties.put("filenameExpression", "payload.name.toUpperCase()");
+		PropertiesInitializer.PROPERTIES = properties;
```
I wonder if we can rely on the `System.properties` or even `ENV` variables...
Do you remember the [12 Factor App](http://12factor.net/) principles ?

artembilan commented on ftp-sink/src/test/java/org/springframework/cloud/stream/module/ftp/FtpSinkIntegrationTests.java:
```diff
@@ -0,0 +1,90 @@
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringApplicationConfiguration(classes = FtpSinkApplication.class, initializers = PropertiesInitializer.class)
+@DirtiesContext
+public class FtpSinkIntegrationTests extends TestFtpServer {
```
This extension smells a bit.
Not sure which goal you aim here, but this architecture looks worse than before.
I see your point regarding `TemporaryFolder` `@Rule`, but maybe we can rework exactly that part and inject `TestFtpServer` as before and use its `getTargetRemoteDirectory()` and so on?..

I won't mind if you go this way anyway...

garyrussell replied:
Maybe it's just the class name that concerns you - the parent class is not really a server; it has a server field. Perhaps it should be called `FtpTestSupport` instead?

artembilan replied:
Yes, `FtpTestServerSupport` sounds good.

ericbottard commented on ftp-sink/src/main/java/org/springframework/cloud/stream/module/ftp/FtpSink.java:
```diff
@@ -0,0 +1,78 @@
+	@Bean
+	public IntegrationFlow ftpInboundFlow() {
+		FtpMessageHandlerSpec handlerSpec =
+					Ftp.outboundAdapter(new FtpRemoteFileTemplate(this.ftpSessionFactory), this.properties.getMode())
```
Indentation seems a bit off, not unreadable though

garyrussell committed:
Rename Test Base Classes

garyrussell commented:
Pushed

ericbottard commented on ftp-sink/pom.xml:
```diff
@@ -0,0 +1,49 @@
+	<description>FTP Sink stream module</description>
+
+	<properties>
+		<start-class>org.springframework.cloud.stream.module.ftp.sink.FtpSinkApplication</start-class>
```
This classname is wrong. Will fix on merge

ericbottard commented:
Merged as 57054a0 as well as https://github.com/spring-cloud/spring-cloud-dataflow/commit/bfd05d9a042d7ad5247f617b85a710f7133bd34e on dataflow

@garyrussell don't forget to add the submodule to the parent pom
